### ISSUE TITLE
JSON fails to parse dictionary keys

### ISSUE THREAD

**Issue by Xrayez** [godotengine/godot#33161]
**Godot version:**
3.2-alpha

**Issue description:**
Converting a variable to JSON with `to_json` method and back with `parse_json` doesn't work on dictionary keys if they are not of type `String`.

**Steps to reproduce:**
```gdscript
extends Node

func _ready():
	var dict = {
		{"key_dict": true}: "value"
	}
	var json = to_json(dict)
	var parsed = parse_json(json)
	
	var key_dict = parsed.keys()[0]
	
	print("Contents: %s, type: %s" % [key_dict, typeof(key_dict)])
	assert(key_dict is Dictionary) # FAIL: got String!
```
Prints:
```
Contents: {key_dict:True}, type: 4
```
**Minimal reproduction project:**
[parse-json-key-dict.zip](https://github.com/godotengine/godot/files/3783978/parse-json-key-dict.zip)

**Comment by Xrayez** [comment]
Hmm, if that's the case then it's a documentation issue, but as dictionary can have arbitrary Variant keys it can get messy serializing stuff as demonstrated in https://github.com/godotengine/godot/pull/33137#issuecomment-547447848. If that's the use case, then `JSON` should not be used, but dictionaries rather serialized with `var2bytes` or similar.

**Comment by Zylann** [comment]
AFAIK Godot already serializes json-incompatible types as string, like vectors.

**Comment by Xrayez** [comment]
@Zylann yeah, it's just that serialization is done implicitly here, see my take at fixing this at #33163.

**Comment by Xrayez** [comment]
Some insights...

Serializing this with `to_json` and back with `parse_json` restores the inner `key_dict` dictionary: 
```gdscript
var dict = {
	"value": {"key_dict": "value"}
}
```
... but not because JSON is able to write and parse Variant types inside it currently, but because `{"key_dict": "value"}` is converted as a string which happens to be another JSON object!

If you take any other JSON-incompatible type for the dictionary *value*, if will also fail to write and parse it (not limited to keys parsing).

So the entire issue can be rephrased as whether JSON should serialize json-incompatible types within  JSON strings, for instance:

```json
{
    "pos": "Vector2( 0, 0 )",
    "color": "Color( 0, 0, 0, 1 )"
}
```
If you take the current implementation and try to serialize it, you'll get this:

```json
{
    "pos": "(0, 0)",
    "color": "0,0,0,1"
}
```
Which can't be converted back to Variant because the JSON writer uses this beast implicitly (instead of `VariantWriter`)

https://github.com/godotengine/godot/blob/d482a1d8e8b4e68f2f386ba618cd091c79c20424/core/variant.cpp#L1411-L1420
```

**Comment by Xrayez** [comment]
Actually this kinda already discussed at #11866, also related to #27529 (talking about inconsistency).

**Comment by KoBeWi** [comment]
Still valid in d98261a
